pkg/k8s/custom_source.go

```diff
@@ -0,0 +1,174 @@
+package k8s
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	// defaultBufferSize is the default number of event notifications that can be buffered.
+	defaultBufferSize = 1024
+)
+
+var _ source.Source = &NotificationChannel{}
+
+// EventType identifies type of PodEvent (CREATE/UPDATE/DELETE)
+type EventType int
+
+const (
+	// CREATE event for given object
+	CREATE EventType = 1
+
+	// DELETE event for given object
+	DELETE EventType = 2
+
+	// UPDATE event for given object
+	UPDATE EventType = 3
+)
+
+// GenericEvent is a wrapper for Create/Update/Delete Events
+type GenericEvent struct {
+	EventType
+
+	// Meta is the ObjectMeta from the incoming request
+	Meta metav1.Object
+
+	// Object is the object from the incoming request
+	Object runtime.Object
+
+	// OldMeta is the ObjectMeta of existing object.Only populated for DELETE and UPDATE requests.
+	OldMeta metav1.Object
+
+	// OldObject is the existing object. Only populated for DELETE and UPDATE requests.
+	OldObject runtime.Object
+}
+
+// NotificationChannel monitors channels of type Create/Update/Delete
+type NotificationChannel struct {
+	// once ensures the event distribution goroutine will be performed only once
+	once sync.Once
+
+	// stop is to end ongoing goroutine, and close the Create channel
+	stop <-chan struct{}
+
+	Source <-chan GenericEvent
```

Comment:
this should be `source`

Reply:
It is used in cloudmap_controller to set the Source. It is an exported field